libyul/YulString.h

```diff
@@ -35,69 +35,84 @@ namespace yul
 class YulStringRepository: boost::noncopyable
 {
 public:
-	YulStringRepository()
+	struct Handle
 	{
-		reset();
-	}
+		size_t id;
+		std::uint64_t hash;
+	};
+	YulStringRepository():
+		m_strings{std::make_shared<std::string>()},
+		m_hashToID{std::make_pair(emptyHash(), 0)}
+	{}
 	static YulStringRepository& instance()
 	{
 		static YulStringRepository inst;
 		return inst;
 	}
-	size_t stringToId(std::string const& _string)
+	Handle stringToHandle(std::string const& _string)
 	{
 		if (_string.empty())
-			return 0;
-		size_t& id = m_ids[_string];
-		if (id == 0)
-		{
-			m_strings.emplace_back(std::make_shared<std::string>(_string));
-			id = m_strings.size() - 1;
-		}
-		return id;
-	}
-	std::string const& idToString(size_t _id) const
-	{
-		return *m_strings.at(_id);
+			return { 0, emptyHash() };
+		std::uint64_t h = hash(_string);
+		auto range = m_hashToID.equal_range(h);
+		for (auto it = range.first; it != range.second; ++it)
+			if (*m_strings[it->second] == _string)
+				return Handle{it->second, h};
+		m_strings.emplace_back(std::make_shared<std::string>(_string));
+		size_t id = m_strings.size() - 1;
```

Comment:
Good idea, but doesn't the allocation of IDs still depend on the insertion order, at least for strings with the same hash?
Also, if we use hashes, why not just let `std::unordered_map<string, id>` handle the hash collisions for us?

Reply:
I still don't fully understand the pros and cons here, let's talk about it later!